File: carcassonne-master/src/field_side_decorator.cc
#include "field_side_decorator.h"

#include "terrain/field_segment.h"

namespace carcassonne
{

FieldSideDecorator::FieldSideDecorator()
{
}

FieldSideDecorator::FieldSideDecorator(Side* side)
	:	TerrainSideDecorator(side)
{
	middle_.reset(new FieldSegment);
	
}

FieldSideDecorator::FieldSideDecorator(const FieldSideDecorator& src) 
	: TerrainSideDecorator(src)
{
	CopyFrom(src);	
}

FieldSideDecorator& FieldSideDecorator::
	operator=(const FieldSideDecorator& rhs)
{
	if(this == &rhs) {
		return (*this);
	}
	
	CopyFrom(rhs);
	
	return (*this);
}
 	
void FieldSideDecorator::
	CopyFrom(const FieldSideDecorator& src)
{
}
 
Side* FieldSideDecorator::
	Clone() const
{
	return new FieldSideDecorator(*this);
}

SideDecorator* FieldSideDecorator::Decorate(Side* side)
{
	return new FieldSideDecorator(side);
}

std::string FieldSideDecorator::ToString() const
{
	return "field " + TerrainSideDecorator::ToString();
	
}

bool FieldSideDecorator::
	IsPotentialTerrainMatch(const TerrainSideDecorator& 
														inTerrainSideDecorator) const
{
	return inTerrainSideDecorator.IsExistingTerrainMatch(*this);
}
 	
bool FieldSideDecorator:: 
	IsExistingTerrainMatch(const FieldSideDecorator& inFieldSideDecorator) const
{
	return true;
}

bool FieldSideDecorator:: 
	IsExistingTerrainMatch(const RoadSideDecorator& inRoadSideDecorator) const
{
	return false;
}

bool FieldSideDecorator::
	IsExistingTerrainMatch(const CitySideDecorator& inCitySideDecorator) const
{
	return false;
}

FieldSideDecorator::~FieldSideDecorator()
{
}

}

File: js/FieldSideDecorator.js
const TerrainSideDecorator=require( "./TerrainSideDecorator");
const Side=require( "./Side");
const RoadSideDecorator=require( "./RoadSideDecorator");
const CitySideDecorator=require( "./CitySideDecorator");
const FieldSegment=require( "./terrain/FieldSegment");
module.exports= class FieldSideDecorator extends TerrainSideDecorator {
    constructor(src) {
        super(src);
        if (src instanceof FieldSideDecorator) {
            this.CopyFrom(src);
        } else if (src instanceof Side) {
            this.middle_ = new FieldSegment();
        }
    }

    Clone() {
        return new FieldSideDecorator(this);
    }

    Decorate(side) {
        return new FieldSideDecorator(side);
    }

    ToString() {
        return "field " + TerrainSideDecorator.ToString();

    }
    IsPotentialTerrainMatch(inTerrainSideDecorator) {
        return inTerrainSideDecorator.IsExistingTerrainMatch(this);
    }



    IsExistingTerrainMatch(inRoadSideDecorator) {
        return inRoadSideDecorator instanceof FieldSideDecorator;
    }
}